### ISSUE TITLE
Auto-complete for Search Input

### ISSUE THREAD

**Issue by mcarrano** [patternfly/patternfly-design#1006]
Support optional auto-complete for the Search Input component. As the user types, the component will supply a list of possible matches allowing the user to pick from the list of available choices. This both reduces the burden on typing and makes it less likely the the user will enter a search term that does not exist. Want to support simple search use cases as well as attribute:value search use cases used in tagging scenarios.

Some design questions:

* For tagging scenarios, does auto-complete apply to values as well as attributes?
* Should this apply to text inputs as well as search inputs? 

KeyCloak use case and design here: https://marvelapp.com/prototype/dhi6e7e/screen/66619908

**Comment by mcarrano** [comment]
@yih-wang I opened this new issue to define the auto-complete behavior associated with your attribute-value search input request (https://github.com/patternfly/patternfly-design/issues/771). I think that we would do something more similar to a Google search where as the user starts typing, a list of potential matches is exposed, like:

![Screen Shot 2021-01-29 at 4 33 32 PM](https://user-images.githubusercontent.com/6619860/106330068-931d2a80-6250-11eb-971c-5cab0f75482b.png)

This is similar to a type-ahead select except that the value list would never be exposed before the user starts typing. For an attribute-value search of the type you describe the component would need to receive a complete set of attributes and/or values as an array of possible matches for this to work.

Let me know if this is consistent with your expectations and we can continue to elaborate.

**Comment by yih-wang** [comment]
@mcarrano This auto-complete makes sense to me. The only question here is I'm not sure whether the search in Keycloak can receive a complete set of attributes and/or values as an array of possible matches. I remember we have some limitations in this respect. @xianli123 will make a confirmation and get back to you.

**Comment by xianli123** [comment]
@mcarrano @yih-wang We've discussed attribute-value search with Keycloak dev team. Type-ahead is a great method for Keycloak. So we will support it. If possible, we can use auto-complete to search for attribute-value in Keycloak new design.
 
Our concern is the deadline. Keycloak has a very tight development schedule. We want to know when this component will be released?

**Comment by mcarrano** [comment]
@xianli123 We expect to have the basic attribute-value search (search input with an integrated search form) available in out next release in approximately two weeks - by Feb 19. But that will not include auto-complete which is something we will need to add later. If I prioritize that, we should be able to get that into a release next month.

**Comment by mcarrano** [comment]
@xianli123 I learned today that the Openshift team has implemented a similar auto-complete search. See the screen shot below. So even if this does not get into PatternFly in your required timeframe, perhaps there is an opportunity for code sharing?

![Screen Shot 2021-02-03 at 3 55 52 PM](https://user-images.githubusercontent.com/6619860/106809940-4b284a00-663a-11eb-9e60-9fb55da2e2a3.png)

If this is something that would help, I can get more information.

**Comment by mcarrano** [comment]
@yih-wang @xianle123 Here is a more fully developed design proposal for this feature. https://docs.google.com/document/d/1mFOuhoAJNBSWKHZKwz8T6OVk71OWNPhfAuVF1Jw3BAk/edit?usp=sharing  Let me know if this will meet you needs. @gdoyle1 perhaps also take a look from an OpenShift perspective?

@jessiehuff I's also love your input regarding keyboard interaction. I think using the Tab key for accepting an auto-complete string is pretty standard, but I'm not 100% sure.

**Comment by jessiehuff** [comment]
I agree, it looks like the Tab key can often be used for accepting an auto-complete string. When researching autocomplete search, it looked like almost all implementations followed a similar or the same pattern which actually matches our Typeahead Select behavior. For this I would expect the following for keyboard interaction:

- UP and DOWN: Cycles hover focus through auto-suggestions and input field.
- ESC: Close the menu
- ENTER: Select the currently focused auto-suggestion and close the menu.
- TAB: Select the currently focused auto-suggestion, close the menu, and move focus to the next focusable element.

There's a number of resources that the developer can refer to on this as well:
https://adamsilver.io/blog/building-an-accessible-autocomplete-control/
https://accessibility.huit.harvard.edu/technique-aria-autocomplete
https://haltersweb.github.io/Accessibility/autocomplete.html?
https://alphagov.github.io/accessible-autocomplete/examples/

**Comment by gdoyle1** [comment]
@mcarrano @yih-wang This would work very well for Openshift!! Design looks good. @bgliwa01 Do you want to check this out for the dev side as well? Made me think about your quick search and if you can have autofill for topology?

**Comment by bgliwa01** [comment]
Yes! I also am doing a design for Pipelines right now that could use this. Another place to look in OpenShift is the metrics tab on metrics tab under monitoring on the admin side. I attached a screenshot 
<img width="1678" alt="Screen Shot 2021-02-16 at 11 28 46 AM" src="https://user-images.githubusercontent.com/51342070/108092059-515ef300-704a-11eb-9dd5-691907e30ab0.png">

**Comment by bgliwa01** [comment]
I talked with Andrew and Christian (OpenShift dev team) and they had some feedback - "So essentially the issue I saw upstream pretty much was “as the user types, show the autocomplete box”. But in reality, we want control over that, so the code around the Input field needs to determine that for our use-case. In this case, when the user hits the right keybinding or when they type '$('. Since those are business logic things, PF can’t write that into their generic component, so they need to provide “API” (for lack of a better word) to us for us to do that. So whatever they may be doing by default (after typing a character or two or whatever) they need to also handle our case. Which they may or may not already overlap, not sure how they’re planning to code it up" and "devs would expect this to autocomplete like an IDE".
